### PR TITLE
TP-695 Edit measure geographical area

### PR DESCRIPTION
**Pull request by DIT-CJWaszczuk**
## Summary
Adds ability to edit the geographical area of a measure 
## Screenshots 
image

### CODE REVIEW

File: geo_areas/models.py [hunk 1]
```diff
@@ -98,6 +98,15 @@ def is_a_parent(self):
             .exists()
         )
 
+    def is_single_region_or_country(self):
+        return self.area_code == 0 or self.area_code == 2
+
+    def is_erga_omnes(self):
```
Comment (stuaxo) [inline geo_areas/models.py:104]: Would be great to have a short docstring saying what erga_omnes is !

File: measures/forms.py [hunk 2]
```diff
@@ -57,10 +58,67 @@ class MeasureForm(ValidityPeriodForm):
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.initial_geographical_area = self.instance.geographical_area
+
+        self.fields[
+            "geographical_area_group"
+        ].label_from_instance = lambda obj: obj.structure_description
+        self.fields[
+            "geographical_area_country_or_region"
+        ].label_from_instance = lambda obj: obj.structure_description
```
Comment (stuaxo) [inline measures/forms.py:91]: I assume these two are meant to be the same ?

File: geo_areas/models.py (hunk 1 above)
Comment (DIT-CJWaszczuk) [reply geo_areas/models.py:104]: I'd rather we just remove the damn phrase all together!

File: measures/forms.py [hunk 3]
```diff
@@ -57,10 +58,67 @@ class MeasureForm(ValidityPeriodForm):
...
+                "geographical_area_country_or_region"
+            ].initial = self.instance.geographical_area
 
     def clean(self):
         cleaned_data = super().clean()
 
+        erga_omnes_instance = GeographicalArea.objects.latest_approved().filter(
+            area_code=1,
+            area_id=1011,
+        )[0]
```
Comment (stuaxo) [inline measures/forms.py:109]: Can the query return multiple objects ?

If it is always one, then `.get(area_code=1, area_id=1011)` may be a better fit than `.filter(....)[0]`.

File: geo_areas/models.py (hunk 1 above)
Comment (stuaxo) [reply geo_areas/models.py:104]: Sounds good, I have no idea what it is... yet another cryptic taric thing.

Comment (simonwo) [reply geo_areas/models.py:104]: Feel free to use `all_countries` instead.

File: geo_areas/models.py [hunk 4]
```diff
@@ -98,6 +98,15 @@ def is_a_parent(self):
             .exists()
         )
 
+    def is_single_region_or_country(self):
+        return self.area_code == 0 or self.area_code == 2
```
Comment (simonwo) [inline geo_areas/models.py:102]: The constants in [`geo_areas.validators.AreaCode`](https://github.com/uktrade/tamato/blob/master/geo_areas/validators.py#L10) would help avoid the magic numbers.

File: measures/forms.py (hunk 3 above)
Comment (simonwo) [reply measures/forms.py:109]: It's technically possible for this [area code to be reused](https://uktrade.github.io/tamato/docs/business_rules/geo_areas.html#geo_areas.business_rules.GA1) which would result in multiple areas with the area ID 1011, so I would advise adding a call to [`as_at(instance.valid_between.lower)`](https://github.com/uktrade/tamato/blob/master/common/models/records.py#L136) here, and then as Stuart said, calling `get`.

**Review by simonwo — CHANGES_REQUESTED** [review]
Can't see anything controversial! 😁 Would be good to land a few tweaks though.

Also did we figure out how to write tests for these things yet?

**Comment by stuaxo** [comment]
> Can't see anything controversial! grin 
Me neither - thought I'd need to reread it as I couldn't see any.

> And I heartily approved of new fields on the model (maybe they could be `@property`?). Would be good to land a few tweaks though.
Had the same thought on property, but wasn't sure I knew the code enough here to write the comment.

> 
> Also did we figure out how to write tests for these things yet?

For the FE ?   Yes, though I need to come back and implement them.

**Comment by stuaxo** [comment]
@andyhd would this make most sense as part of the QuerySet ?

File: measures/forms.py (hunk 2 above)
Comment (DIT-CJWaszczuk) [reply measures/forms.py:91]: yeah i'm concerned of the select populating with an item that shouldn't be in, i.e you previously had turkey as the area, you didn't want Turkey showing in the groups, can probably clean it up

File: measures/forms.py (hunk 3 above)
Comment (stuaxo) [reply measures/forms.py:109]: @andyhd would erga_omnes make most sense as part of the QuerySet ?

So this ends up with something like:
GeographicalArea.objects.erga_omnes.latest_approved()

File: measures/forms.py [hunk 5]
```diff
@@ -57,10 +59,108 @@ class MeasureForm(ValidityPeriodForm):
...
+        self.fields[
+            "additional_code"
+        ].queryset = AdditionalCode.objects.approved_up_to_transaction(tx)
+        self.fields[
+            "geographical_area"
+        ].queryset = GeographicalArea.objects.approved_up_to_transaction(tx)
+        self.fields[
+            "geographical_area_group"
+        ].queryset = GeographicalArea.objects.approved_up_to_transaction(tx).filter(
+            area_code=1,
```
Comment (andyhd) [inline measures/forms.py:116]: I wonder about the order of `approved_up_to_transaction` and `filter` - does it matter which way round they go? I haven't looked into the details, but I feel like filtering by `area_code` first will mean the `approved_up_to_transaction` will have fewer rows to operate on and will therefore be quicker? @stuaxo any thoughts?

**Review by andyhd — APPROVED** [review]
Looks good to me 👍

**Pull request merged**